File: mdqt/source/qt/getLaserPolarization.cpp
#include "qt.h"

// This function outputs the spherical tensor components of the LIF-laser polarization
// when projected into the local quantization axis set by the quadrupole magnetic field model.

using namespace std;

void QT::getLaserPolarization(complex<double> &epsM,complex<double> &eps0,complex<double> &epsP) const
// Note: 'option' allows the user to select the laser's polarization as defined in the lab frame, and is not
//          directory indicative of the laser polarization as expressed in local magnetic field basis.

// eta: dimensionless parameter ranging from 0 to 1, sets ratio between y- and z-components of laser polarization.
//      (eta = 1 --> y-polarization only) and (eta = 0 --> z-polarization only)
//      (eta = 1/sqrt(2) -> circular polarization)
// xi: phase shift of y-component of polarization in radians.
//      (xi = -pi/2 -> left-handed polarization)
//      (xi = pi/2 -> right-handed polarization)
// This function also uses the 'theta' parameter from the QT class, which specifies the angle (in radians) that
// the y-axis of the lab-frame coordinate system makes with the local magnetic field vector.

// eps[i]: i = 0
{
    // choose eta and xi based on user-defined 'option'
    double eta;
    double xi;
    if (pol.compare("lin") == 0 && B != 0){
        eta = 1.;
        xi = 0.;
    }
    else if (pol.compare("right") == 0 && B != 0){
        eta = 1./sqrt(2.);
        xi = -cts::pi/2.;
    }
    else if (pol.compare("left") == 0 && B != 0){
        eta = 1./sqrt(2.);
        xi = cts::pi/2.;
    }
    
    // define laser polarization in spherical tensor basis
    // eps = [eps0 epsP epsM]
    complex<double> i(0.,1.);
    eps0 = eta*cos(theta)*cos(phi)+exp(i*xi)*sqrt(1.-pow(eta,2.))*sin(phi);
    epsP = 1./sqrt(2.)*(-exp(i*xi)*sqrt(1.-pow(eta,2.))*cos(phi)-i*eta*sin(theta)+eta*cos(theta)*sin(phi));
    epsM = 1./sqrt(2.)*(+exp(i*xi)*sqrt(1.-pow(eta,2.))*cos(phi)-i*eta*sin(theta)-eta*cos(theta)*sin(phi));
    
    // if the magnetic field is zero, then the quantization axis is set by the LIF-laser
    if (B==0){
        if (pol.compare("lin")==0){
            epsP = 0;
            epsM = 0;
            eps0 = 1;
        }
        else if (pol.compare("left")==0){
            eps0 = 0;
            epsM = 1;
            epsP = 0;
        }
        else if (pol.compare("right")==0){
            eps0 = 0;
            epsM = 0;
            epsP = 1;
        }
    }
}

File: spectrum-model/source/getLaserPolarization.m
function [eps,epsM,eps0,epsP] = getLaserPolarization(pol,theta,phi,b,B)
% pol (string): LIF-laser polarization - 'lin','left','right'
% theta (double): angle (radians) between local magnetic field and y-axis of lab frame
% phi (double): angle (radians) between local field and x-y plane
% b (struct): contains information about basis states - see 'defineBasisStates.m'
% B (double): magnitude of local magnetic field in Tesla

%% Error Checking
% ensure 'pol' is specified correctly
if max(strcmp(pol,{'left','right','lin'})) ~= 1
    error('<pol> specified incorrectly. String options are ''lin'', ''left'', ''right''')
end

%% Function
% the following equations will be written up in Grant Gorman's thesis. For the phi = 0 equations, see the magnetized LIF
% paper.

% define polarization in spherical tensor notation for laser polarization vector in local field frame
i = sqrt(-1);
e0 = @(eta,xi,theta,phi) eta*cos(theta)*cos(phi)+exp(i*xi)*sqrt(1-eta^2)*sin(phi); % dm = 0
eP = @(eta,xi,theta,phi) 1/sqrt(2)*(-exp(i*xi)*sqrt(1-eta^2)*cos(phi)-i*eta*sin(theta)+eta*cos(theta)*sin(phi)); % dm = +1
eM = @(eta,xi,theta,phi) 1/sqrt(2)*(+exp(i*xi)*sqrt(1-eta^2)*cos(phi)-i*eta*sin(theta)-eta*cos(theta)*sin(phi)); % dm = -1

% based on user input of 'pol' and quantization axis, choose 'eta' and 'xi' 
if strcmp(pol,'right') && B~= 0   % for right-handed circular polarization, y = -pi/2
    eta = 1/sqrt(2);
    xi = -pi/2;
    eps0 = e0(eta,xi,theta,phi);
    epsP = eP(eta,xi,theta,phi);
    epsM = eM(eta,xi,theta,phi);
elseif strcmp(pol,'left')  && B~= 0   % for left-handed circular polarization, y = pi/2
    eta = 1/sqrt(2);
    xi = +pi/2;
    eps0 = e0(eta,xi,theta,phi);
    epsP = eP(eta,xi,theta,phi);
    epsM = eM(eta,xi,theta,phi);
elseif strcmp(pol,'lin')  && B~= 0 % linear laser polarization
    eta = 1;
    xi = 0;
    eps0 = e0(eta,xi,theta,phi);
    epsP = eP(eta,xi,theta,phi);
    epsM = eM(eta,xi,theta,phi);
elseif strcmp(pol,'lin') && B == 0 % laser sets quantization axis in absense of local fields
    epsP = 0;
    epsM = 0;
    eps0 = 1;
elseif strcmp(pol,'left') && B == 0 % laser sets quantization axis in absense of local fields: theta = pi/2
    eps0 = 0;
    epsM = 1;
    epsP = 0;
elseif strcmp(pol,'right') && B == 0 % laser sets quantization axis in absense of local fields: theta = pi/2
    eps0 = 0;
    epsM = 0;
    epsP = 1;
end

if isnan(eps0*epsP*epsM)
    error('Spherical tensor components of LIF-laser polarization are not valid. Check user input.')
end

% define eps(i,j) the component of laser polarization aligned with transition |i> to |j>
eps = zeros(length(b.s));
for i = b.exc % final state
    for j = b.gnd % initial state
        dm = b.m(i) - b.m(j);
        if i ~= j
            if dm == 0
                eps(i,j) = eps0;
            elseif dm == 1
                eps(i,j) = epsP;
            elseif dm == -1
                eps(i,j) = epsM;
            end
            eps(j,i) = eps(i,j);
        end
    end
end


end